### PR TITLE
Catch exceptions in control blocks and fail the control

### LINKED ISSUE #2981: Exceptions are not properly caught by control blocks
### 🎛 Description

🐛 bug report

In some circumstances, if an exception is thrown in a `control` block (rather than a `describe` block), the exception is not caught, and the inspec run aborts with stacktrace.  Current doctrine is that no exception in the profile code is allowed to derail the run; rather an exception in a control should be caught and the control marked failed.

Also check to see if this is true for inspec check, inspec exec, or both.

Appears to be aggravated by https://github.com/chef/inspec/pull/2359

refs @chris-rock 

### 🌍 InSpec and Platform Version

2.1.54

### 🤔 Replication Case

Full repro profile available on branch https://github.com/chef/inspec/tree/repro-2981

Example (Explodes on MacOS with no sshd)

```
# sshd05 throws exception which is not caught, and aborts inspec with stacktrace
# Defective on 2.1.54
control 'sshd05 sshd_config resource nonexistant path with contents access control block' do
  s = sshd_config('/i/do/not/exist')
  # Next line triggers compile-time exception
  c = s.Protocol
  describe 'Test block' do
    subject { c }
    it { should be_nil }
  end
end
```

### 💁 Possible Solutions

Meditate on describe and InSpec's extensions to it.

### 💻 Stacktrace

```
[cwolfe@lodi inspec-aux]$ be inspec exec test/regression/2981-exceptions-in-control-blocks/profile-01/
bundler: failed to load command: inspec (/Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/bin/inspec)
Inspec::Exceptions::ResourceSkipped: Can't find file: /i/do/not/exist
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/utils/file_reader.rb:11:in `read_file_content'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/resources/ssh_conf.rb:64:in `read_content'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/resources/ssh_conf.rb:69:in `read_params'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/resources/ssh_conf.rb:48:in `method_missing'
  test/regression/2981-exceptions-in-control-blocks/profile-01/controls/02-sshd_config.rb:41:in `block in load_with_context'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/rule.rb:50:in `instance_eval'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/rule.rb:50:in `initialize'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/control_eval_context.rb:75:in `new'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/control_eval_context.rb:75:in `block (2 levels) in create'
  test/regression/2981-exceptions-in-control-blocks/profile-01/controls/02-sshd_config.rb:38:in `load_with_context'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile_context.rb:158:in `instance_eval'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile_context.rb:158:in `load_with_context'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile_context.rb:142:in `load_control_file'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile.rb:170:in `block in collect_tests'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile.rb:167:in `each'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/profile.rb:167:in `collect_tests'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/runner.rb:93:in `block in load'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/runner.rb:82:in `each'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/runner.rb:82:in `load'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/runner.rb:103:in `run'
  /Users/cwolfe/sandbox/inspec/inspec-aux/lib/inspec/cli.rb:168:in `exec'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
  /Users/cwolfe/sandbox/inspec/inspec-aux/bin/inspec:12:in `<top (required)>'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/bin/inspec:23:in `load'
  /Users/cwolfe/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/bin/inspec:23:in `<top (required)>'
```

### PR DESCRIPTION
**Pull request by clintoncwolfe**
When a control block has any kind of exception thrown, we currently abort the run.  If it's an Inspec exception, we (sometimes regrettably) suppress the stacktrace and exit with code 1.  Other exceptions explode in the normal way.

This behavior goes again doctrine - that nothing in the profile code can derail the inspec run.  In this PR, we catch the exception as early as possible; then to fail the control, we inject a dummy resource constructed so that the source location and exception error are preserved.

A functional test has been added to verify the new behavior.  Additionally, a document has been added for future developers outlining how profile code is evaluated.

For special review: is the messaging produced when a control fails useful?  Could it be reworded, or should we add some of the stack trace or other information available in the caught exception?

```
[cwolfe@lodi inspec-aux]$ be inspec exec test/unit/mock/profiles/exception-in-control

Profile: InSpec Profile (profile-01)
Version: 0.1.0
Target:  local://

  ✔  sshd01 using sshd_config resource nonexistant path with no property access: Test block
     ✔  Test block should include "Test"
  ×  sshd02 sshd_config resource nonexistant path with contents access in test block: Protocol
     ×  Test block Protocol
     Can't find file: /i/do/not/exist
  ✔  sshd03 sshd_config resource nil path with contents access in test block: Protocol
     ✔  Test block Protocol should be nil
  ✔  sshd04 sshd_config resource default path with contents access in test block: Protocol
     ✔  Test block Protocol should be nil
  ×  sshd05 sshd_config resource nonexistant path with contents access control block: test/unit/mock/profiles/exception-in-control/controls/02-sshd_config.rb:38
     ×  Control Source Code Error test/unit/mock/profiles/exception-in-control/controls/02-sshd_config.rb:38
     Can't find file: /i/do/not/exist
  ✔  c01 using file resource on nonexistant file with no property access: Test block
     ✔  Test block should include "Test"
  ✔  c02 using file resource on nonexistant file with contents access in test block: Test block
     ✔  Test block should be nil
  ✔  c03 using file resource on nonexistant file with contents access control block: Test block
     ✔  Test block should be nil
  ×  a control with an undefined method: test/unit/mock/profiles/exception-in-control/controls/03-syntax.rb:3
     ×  Control Source Code Error test/unit/mock/profiles/exception-in-control/controls/03-syntax.rb:3
     undefined local variable or method `swords' for #<#<Class:0x00007fb0dc12fa18>:0x00007fb0df6f1b60>
  ×  a control with a divide-by-zero: test/unit/mock/profiles/exception-in-control/controls/03-syntax.rb:7
     ×  Control Source Code Error test/unit/mock/profiles/exception-in-control/controls/03-syntax.rb:7
     divided by 0


Profile Summary: 6 successful controls, 4 control failures, 0 controls skipped
Test Summary: 6 successful, 4 failures, 0 skipped
[cwolfe@lodi inspec-aux]$
```

Fixes #2981

### CODE REVIEW

**Review by jquick — APPROVED** [review]
Fantastic work @clintoncwolfe. Thanks for the excellent documentation also!

File: docs/dev/control-eval.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    #5  Inspec::Profile.collect_tests(include_list#Array) at lib/inspec/profile.rb:167
+    #4  Hash.each at lib/inspec/profile.rb:167
+    #3  block in Inspec::Profile.block in collect_tests(include_list#Array) at lib/inspec/profile.rb:170
+    #2  Inspec::ProfileContext.load_control_file(*args#Array) at lib/inspec/profile_context.rb:141
+    #1  Inspec::ProfileContext.control_eval_context at lib/inspec/profile_context.rb:58
+    #0  #<Class:Inspec::ControlEvalContext>.create(profile_context#Inspec::ProfileContext, resources_dsl#Module) at lib/inspec/control_eval_context.rb:41
+
+## A profile context is created
+
+Like many things in InSpec core, a profile context is an anonymous class. (verify)
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:23]: Could we remove `(verify)`?

File: docs/dev/control-eval.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+Additionally, a control_eval_context is created.  It is an instance of an anonymous class; it has a class<->relationship with its profile context.  See `lib/inspec/control_eval_context.rb`.
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:25]: I think we could do a reference link to that file if you so chose. I don't mind either way.

File: docs/dev/control-eval.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+## Each file's contents are instance eval'd against the control_eval_context
+
+### DSL methods are executed at this time
+
+So, if you have a control file with `title` in it, that will call the title method that was defined at `lib/inspec/control_eval_context.rb:60`.  Importantly, this also includes the `control` DSL keyword, and also the `describe` keyword (used for bare describes).
+
+### Each control and their block are wrapped in an anonymous class
+
+The anonymous class generator is located at `lib/inspec/control_eval_context.rb:24`.  At this point, the terminology switches from `control` to `rule`.  Each context class inherits from Inspec::Rule, which provides the constructor.
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:35]: Same as above, link is nice but not required.

File: docs/dev/control-eval.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+The control context class also gets extended with the resource DSL, so anything in the source code for the control can use the resource DSL.  This includes all resource names, but importantly, the `describe` DSL keyword.
+
+Finally, Inspec::Rule provides the control DSL - impact, title, desc, ref, and tags.
+
+### The block is instance_eval'd against the control context class
+
+See `lib/inspec/rule.rb:50`.  We're now in two levels of instance eval'ing - the file is gradually being eval'd against the profile context anonymous class, and the current control's block is being instance eval'd against a control context anonymous class.
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:43]: Another potential link

File: docs/dev/control-eval.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+At this stage, control-level metadata (impact, title, refs, tags, desc) are evaluated and set as instance vars on the control.
+
+Any "loose" ruby in the control is also executed at this point.
+
+And, the describe and describe.one blocks are executed.
+
+### TODO: describe blocks are *not* instance-evaled
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:51]: Should we remove the TODO?

File: docs/dev/control-eval.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+### The control is registered with the profile
+
+Using the method register_control (dynamically defined on the control eval context), we check for various skip conditions.  If none of them apply, the control is then registered with the profile context using register_rule.
+
+ProfileContext.register_rule's main job is to determine the full ID of the control (within the context of the profile) and either add it to the controls list, or (if another control with the same ID exists), merge it.  (This is where overriding happens).
+
+Note: can skip a control with:
+Inspec::Rule.set_skip_rule(control, msg)
+
+## What else?
```
Comment (jerryaldrichiii) [inline docs/dev/control-eval.md:62]: Could we drop this line?

File: test/unit/mock/profiles/exception-in-control/inspec.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+name: profile-01
+title: InSpec Profile
+maintainer: The Authors
```
Comment (jerryaldrichiii) [inline test/unit/mock/profiles/exception-in-control/inspec.yml:3]: Some of the other `inspec.yml` (all others?) use different metadata values. Not sure if it important though.

See: https://github.com/chef/inspec/blob/master/test/unit/mock/profiles/aws-profile/inspec.yml

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
Looks great! I've added some comments. As usual 100% nitpicks. Feel free to disregard most.

File: test/unit/mock/profiles/exception-in-control/inspec.yml (hunk 7 above)
Comment (clintoncwolfe) [reply test/unit/mock/profiles/exception-in-control/inspec.yml:3]: I just used whatever `inspec init profile` generated.  Nothing in the metadata was relevant to the test underway.

File: docs/dev/control-eval.md (hunk 1 above)
Comment (clintoncwolfe) [reply docs/dev/control-eval.md:23]: This is a document for us, the inspec team.  I wouldn't expect it to be consumed publicly, or if so, only by people trying to work on the guts of inspec. Much lower standdards - the goal is to capture tribal knowledge.  It's incomplete, both structurally and content-wise.  Some areas might even be factually incorrect.  Here, I'm just tempering my certainty; the particular snipe-hunt I was on didn't entail me looking very hard at the profile context.  Next time someone needs to know this sort of tribal knowledge, perhaps they can shore it up a bit more, if their research leads them in that direction.

**Review by miah — APPROVED** [review]
Excellent documentation. =)

File: docs/dev/control-eval.md (hunk 1 above)
Comment (jerryaldrichiii) [reply docs/dev/control-eval.md:23]: Ah, that makes sense. I suppose leaving those sorts of things in are fine then.

File: test/unit/mock/profiles/exception-in-control/inspec.yml (hunk 7 above)
Comment (jerryaldrichiii) [reply test/unit/mock/profiles/exception-in-control/inspec.yml:3]: I figured, I'll mark it as approved, I don't think the copyright lines and such really matter. Was just going for consistency.

**Comment by clintoncwolfe** [comment]
Overriding Jerry's review because he said it was all nitpicks, and he was OK with me overriding.

**Review by jerryaldrichiii — APPROVED** [review]
Thanks for all the work @clintoncwolfe. Especially the documentation. It's always needed!

**Pull request merged**